--- shibboleth.spec.org	2022-01-13 03:43:51.897716685 +0000
+++ shibboleth.spec	2022-01-13 03:47:13.573580156 +0000
@@ -7,9 +7,13 @@
 License:	Apache-2.0
 URL:		http://shibboleth.net/
 Source:		%{name}-sp-%{version}.tar.bz2
+Patch0:		logger.patch
+Patch1:		shibresponder.patch
 BuildRoot:	%{_tmppath}/%{name}-sp-%{version}-root
 Obsoletes:	shibboleth-sp = 2.5.0
 Requires:	openssl
+Requires: 	supervisor
+Requires: 	nginx
 %if 0%{?rhel} >= 6 || 0%{?centos_version} >= 600 || 0%{?amzn} >= 1
 PreReq:		xmltooling-schemas%{?_isa} >= 3.0.0, opensaml-schemas%{?_isa} >= 3.0.0
 %else
@@ -18,19 +22,16 @@
 %if 0%{?suse_version} > 1030 && 0%{?suse_version} < 1130
 PreReq:		%{insserv_prereq} %{fillup_prereq}
 %endif
-%if 0%{?rhel} >= 7
+%if 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 BuildRequires:  systemd-devel
 %endif
-%if 0%{?rhel} >= 8
-BuildRequires: gdb
-%endif
 BuildRequires:  libxerces-c-devel >= 3.2
 BuildRequires:	libxml-security-c-devel >= 2.0.0
 BuildRequires:	libxmltooling-devel >= 3.0.0
 BuildRequires:	libsaml-devel >= 3.0.0
 %{?_with_log4cpp:BuildRequires: liblog4cpp-devel >= 1.0}
 %{!?_with_log4cpp:BuildRequires: liblog4shib-devel >= 2}
-%if 0%{?rhel} == 6 || 0%{?rhel} == 7 || 0%{?amzn} >= 1
+%if 0%{?rhel} == 6 || 0%{?rhel} == 7 || 0%{?amzn} >= 1
 Requires:	libcurl-openssl%{?_isa} >= 7.21.7
 BuildRequires:	chrpath
 %endif
@@ -39,17 +40,15 @@
 %endif
 BuildRequires:  gcc-c++, pkgconfig, boost-devel >= 1.32.0
 %{!?_without_gssapi:BuildRequires: krb5-devel}
-%if 0%{?rhel} < 8
 %{!?_without_doxygen:BuildRequires: doxygen}
-%endif
 %{!?_without_odbc:BuildRequires:unixODBC-devel}
 %{?_with_fastcgi:BuildRequires: fcgi-devel}
-%if 0%{?centos} == 6 || 0%{?centos} == 7
+%if 0%{?centos_version} >= 600
 BuildRequires:	libmemcached-devel
 %endif
 %{?_with_memcached:BuildRequires: libmemcached-devel}
 %if "%{_vendor}" == "redhat" || "%{_vendor}" == "amazon"
-%if 0%{?rhel} >= 6 || 0%{?amzn} >= 1
+%if 0%{?rhel} >= 6 || 0%{?centos_version} >= 600 || 0%{?amzn} >= 1
 %{!?_without_builtinapache:BuildRequires: httpd-devel%{?_isa}}
 %else
 %{!?_without_builtinapache:BuildRequires: httpd-devel}
@@ -106,6 +105,8 @@
 
 %prep
 %setup -n %{name}-sp-%{version}
+%patch0 -p1
+#%patch1 -p1
 
 %build
 %if 0%{?suse_version} >= 1300
@@ -114,13 +115,10 @@
 %if 0%{?suse_version} >= 1210
 	%configure %{?_without_odbc:--disable-odbc} %{?_without_adfs:--disable-adfs} %{?_with_fastcgi} %{!?_without_gssapi:--with-gssapi} %{!?_without_systemd:--enable-systemd} %{?shib_options}
 %else
-%if 0%{?rhel} >= 8
-	%configure %{?_without_odbc:--disable-odbc} %{?_without_adfs:--disable-adfs} %{?_with_fastcgi} %{!?_without_gssapi:--with-gssapi} %{?_with-memcached} %{!?_without_systemd:--enable-systemd} %{?shib_options}
-%else
-%if 0%{?rhel} >= 7
+%if 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 	%configure %{?_without_odbc:--disable-odbc} %{?_without_adfs:--disable-adfs} %{?_with_fastcgi} %{!?_without_gssapi:--with-gssapi} %{!?_without_memcached:--with-memcached} %{!?_without_systemd:--enable-systemd} %{?shib_options} PKG_CONFIG_PATH=/opt/shibboleth/%{_lib}/pkgconfig
 %else
-%if 0%{?centos} >= 6
+%if 0%{?centos_version} >= 600
 	%configure %{?_without_odbc:--disable-odbc} %{?_without_adfs:--disable-adfs} %{?_with_fastcgi} %{!?_without_gssapi:--with-gssapi} %{!?_without_memcached:--with-memcached} %{?shib_options} PKG_CONFIG_PATH=/opt/shibboleth/%{_lib}/pkgconfig:./pkgconfig-workarounds/rh6
 %else
 %if 0%{?rhel} >= 6
@@ -136,7 +134,6 @@
 %endif
 %endif
 %endif
-%endif
 %{__make} pkgdocdir=%{pkgdocdir}
 
 %install
@@ -173,17 +170,21 @@
 	fi
 	if [ "$APACHE_CONFD" != "no" ] ; then
 		%{__mkdir} -p $RPM_BUILD_ROOT$APACHE_CONFD
-		%{__cp} -p $RPM_BUILD_ROOT%{_sysconfdir}/shibboleth/$APACHE_CONFIG $RPM_BUILD_ROOT$APACHE_CONFD/shib.conf 
+		%{__cp} -p $RPM_BUILD_ROOT%{_sysconfdir}/shibboleth/$APACHE_CONFIG $RPM_BUILD_ROOT$APACHE_CONFD/shib.conf
 		echo "%config(noreplace) $APACHE_CONFD/shib.conf" >> rpm.filelist
 	fi
 fi
 
 # Establish location of systemd file, if any.
 SYSTEMD_SHIBD="no"
-%if 0%{?suse_version} >= 1210 || 0%{?rhel} >= 7
+%if 0%{?suse_version} >= 1210 || 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 	%{__mkdir} -p $RPM_BUILD_ROOT%{_unitdir}
 	echo "%attr(0444,-,-) %{_unitdir}/shibd.service" >> rpm.filelist
 	SYSTEMD_SHIBD="$RPM_BUILD_ROOT%{_unitdir}/shibd.service"
+	echo "%attr(0444,-,-) %{_unitdir}/shibfcgi.service" >> rpm.filelist
+	SYSTEMD_SHIBFCGI="$RPM_BUILD_ROOT%{_unitdir}/shibfcgi.service"
+	echo "%attr(0444,-,-) %{_sysconfdir}/shibboleth/shibfcgi-supervisor.conf" >> rpm.filelist
+	SHIBFCGI_SUPERVISOR_CONF="$RPM_BUILD_ROOT%{_sysconfdir}/shibboleth/shibfcgi-supervisor.conf"
 
 	# Get run directory created at boot time.
 	%{__mkdir} -p $RPM_BUILD_ROOT%{_tmpfilesdir}
@@ -214,20 +215,16 @@
 [Unit]
 Description=Shibboleth Service Provider Daemon
 Documentation=https://wiki.shibboleth.net/confluence/display/SP3/Home
-After=network.target
-Before=httpd.service
+After=network-online.target
 
 [Service]
 Type=notify
 NotifyAccess=main
 User=%{runuser}
-%if 0%{?rhel} == 6 || 0%{?rhel} == 7 || 0%{?amzn} >= 1
+%if 0%{?rhel} == 6 || 0%{?rhel} == 7 || 0%{?amzn} >= 1
 Environment=LD_LIBRARY_PATH=/opt/shibboleth/%{_lib}
 %endif
 ExecStart=%{_sbindir}/shibd -f -F
-StandardInput=null
-StandardOutput=null
-StandardError=journal
 TimeoutStopSec=1m
 TimeoutStartSec=5m
 Restart=on-failure
@@ -236,6 +233,40 @@
 [Install]
 WantedBy=multi-user.target
 EOF
+  cat > $SYSTEMD_SHIBFCGI <<EOF
+[Unit]
+Description=Shibboleth Fast CGI support service
+
+[Service]
+Type=forking
+ExecStart=/usr/bin/supervisord -c %{_sysconfdir}/shibboleth/shibfcgi-supervisor.conf
+
+[Install]
+WantedBy=multi-user.target
+EOF
+  cat > $SHIBFCGI_SUPERVISOR_CONF <<EOF
+[supervisord]
+logfile=%{_localstatedir}/log/shibboleth/supervisord.log
+
+[fcgi-program:shibauthorizer]
+command=%{_libdir}/shibboleth/shibauthorizer
+socket=unix://%{_localstatedir}/run/shibboleth/shibauthorizer.sock
+socket_owner=shibd:shibd
+socket_mode=0660
+user=shibd
+process_name=%%(program_name)s_%%(process_num)02d
+stdout_logfile=%{_localstatedir}/log/shibboleth/shibauthorizer.log
+stderr_logfile=%{_localstatedir}/log/shibboleth/shibauthorizer.error.log
+
+[fcgi-program:shibresponder]
+command=%{_libdir}/shibboleth/shibresponder
+socket=unix://%{_localstatedir}/run/shibboleth/shibresponder.sock
+socket_owner=shibd:shibd
+socket_mode=0660
+user=shibd
+stdout_logfile=%{_localstatedir}/log/shibboleth/shibresponder.log
+stderr_logfile=%{_localstatedir}/log/shibboleth/shibresponder.error.log
+EOF
 elif [ "$SYSCONFIG_SHIBD" != "no" ] ; then
 	# Populate the sysconfig file.
 	cat > $SYSCONFIG_SHIBD <<EOF
@@ -250,7 +281,7 @@
 # Wait period (secs) for configuration (and metadata) to load
 SHIBD_WAIT=30
 EOF
-	%if 0%{?rhel} == 6 || 0%{?rhel} == 7 || 0%{?amzn} >= 1
+	%if 0%{?rhel} >= 6 || 0%{?centos_version} >= 600 || 0%{?amzn} >= 1
 		cat >> $SYSCONFIG_SHIBD <<EOF
 
 # Override OS-supplied libcurl
@@ -259,7 +290,7 @@
 	%endif
 fi
 
-%if 0%{?rhel} == 6 || 0%{?rhel} == 7 || 0%{?amzn} >= 1
+%if 0%{?rhel} >= 6 || 0%{?centos_version} >= 600 || 0%{?amzn} >= 1
 	# Strip existing rpath to libcurl.
 	chrpath -d $RPM_BUILD_ROOT%{_sbindir}/shibd
 	chrpath -d $RPM_BUILD_ROOT%{_bindir}/mdquery
@@ -290,11 +321,14 @@
 	-d  %{_localstatedir}/run/shibboleth -s /sbin/nologin -c "Shibboleth SP daemon" %{runuser}
 %if 0%{?suse_version} >= 1210
 	%service_add_pre shibd.service
+	%service_add_pre shibfcgi.service
 %endif
 exit 0
 
 %post
+%ifnos solaris2.8 solaris2.9 solaris2.10 solaris2.11
 /sbin/ldconfig
+%endif
 
 # Key ownership fix.
 if [ -f %{_sysconfdir}/shibboleth/sp-key.pem ] ; then
@@ -331,9 +365,13 @@
 		fi
 	fi
 
-%if 0%{?rhel} >= 7
+%if 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 	# Initial prep for systemd
-	%systemd_post shibd.service
+  systemctl enable shibd.service
+  systemctl enable shibfcgi.service
+	if [ $1 -eq 1 ] ; then
+		gpasswd -a nginx shibd
+	fi
 	if [ $1 -gt 1 ] ; then
 		systemctl daemon-reload
 	fi
@@ -345,6 +383,7 @@
 %if "%{_vendor}" == "suse"
 %if 0%{?suse_version} >= 1210
 	%service_add_post shibd.service
+	%service_add_post shibfcgi.service
 	systemd-tmpfiles --create %{_tmpfilesdir}/%{name}.conf
 %else
 	# This adds the proper /etc/rc*.d links for the script
@@ -358,8 +397,12 @@
 %preun
 # On final removal, stop shibd and remove service, restart Apache if running.
 %if "%{_vendor}" == "redhat" || "%{_vendor}" == "amazon"
-%if 0%{?rhel} >= 7
+%if 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 	%systemd_preun shibd.service
+	%systemd_preun shibfcgi.service
+	if [ $1 -eq 0 ] ; then
+		gpasswd -d nginx shibd
+	fi
 %else
 	if [ $1 -eq 0 ] ; then
 		/sbin/service shibd stop >/dev/null 2>&1
@@ -373,6 +416,7 @@
 %if "%{_vendor}" == "suse"
 %if 0%{?suse_version} >= 1210
         %service_del_preun shibd.service
+        %service_del_preun shibfcgi.service
 %else
 	%stop_on_removal shibd
 %endif
@@ -383,11 +427,14 @@
 exit 0
 
 %postun
+%ifnos solaris2.8 solaris2.9 solaris2.10 solaris2.11
 /sbin/ldconfig
+%endif
 %if "%{_vendor}" == "redhat" || "%{_vendor}" == "amazon"
 # On upgrade, restart components if they're already running.
-%if 0%{?rhel} >= 7
+%if 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 	%systemd_postun_with_restart shibd.service
+	%systemd_postun_with_restart shibfcgi.service
 %else
 	if [ $1 -ge 1 ] ; then
 		/sbin/service shibd status 1>/dev/null && /sbin/service shibd restart 1>/dev/null
@@ -402,7 +449,7 @@
 %if 0%{?suse_version} >= 1210
 	%service_del_postun shibd.service
 %else
-	cd / 
+	cd /
 	%restart_on_update shibd
 	%{insserv_cleanup}
 %endif
@@ -430,9 +477,9 @@
 %exclude %{_libdir}/*.la
 %dir %{_libdir}/shibboleth
 %{_libdir}/shibboleth/*.so
+%{_libdir}/shibboleth/shibauthorizer
+%{_libdir}/shibboleth/shibresponder
 %exclude %{_libdir}/shibboleth/*.la
-%{?_with_fastcgi:%{_libdir}/shibboleth/shibauthorizer}
-%{?_with_fastcgi:%{_libdir}/shibboleth/shibresponder} 
 %attr(0750,%{runuser},%{runuser}) %dir %{_localstatedir}/log/shibboleth
 %if 0%{?suse_version} < 1300
 %attr(0755,%{runuser},%{runuser}) %dir %{_localstatedir}/run/shibboleth
@@ -452,7 +499,7 @@
 %config(noreplace) %{_sysconfdir}/shibboleth/*.html
 %config(noreplace) %{_sysconfdir}/shibboleth/*.logger
 %if "%{_vendor}" == "redhat"
-%if 0%{?rhel} >= 7
+%if 0%{?rhel} >= 7 || 0%{?centos_version} >= 700
 %else
 %config %{_initrddir}/shibd
 %endif
@@ -464,9 +511,6 @@
 %config %{_initrddir}/shibd
 %{_sbindir}/rcshibd
 %endif
-%if 0%{?suse_version} >= 1210 || 0%{?rhel} >= 7
-%{_tmpfilesdir}/%{name}.conf
-%endif
 %{_sysconfdir}/shibboleth/example-shibboleth2.xml
 %{_sysconfdir}/shibboleth/*.dist
 %{_sysconfdir}/shibboleth/apache*.config
@@ -475,9 +519,7 @@
 %attr(0755,root,root) %{_sysconfdir}/shibboleth/metagen.sh
 %attr(0755,root,root) %{_sysconfdir}/shibboleth/seckeygen.sh
 %doc %{pkgdocdir}
-%if 0%{?rhel} < 8
 %exclude %{pkgdocdir}/api
-%endif
 
 %files devel
 %defattr(-,root,root,-)
@@ -485,14 +527,9 @@
 %{_libdir}/libshibsp.so
 %{_libdir}/libshibsp-lite.so
 %{_libdir}/pkgconfig/*.pc
-%if 0%{?rhel} < 8
 %doc %{pkgdocdir}/api
-%endif
 
 %changelog
-* Mon Sep 30 2019 Scott Cantor <cantor.2@osu.edu> - 3.0.4-1
-- CentOS 8 cleanup
-
 * Mon Apr 30 2018 Scott Cantor <cantor.2@osu.edu> - 3.0.0-1
 - Bump dependency versions
 - Require updated libraries across the board
@@ -548,7 +585,7 @@
 * Sun Jun 26 2011  Scott Cantor  <cantor.2@osu.edu>  - 2.4.3-1
 - Log files shouldn't be world readable.
 - Explicit requirement for libcurl-openssl on RHEL6
-- Uncomment LD_LIBRARY_PATH in init script for RHEL6 
+- Uncomment LD_LIBRARY_PATH in init script for RHEL6
 - Remove rpath from binaries for RHEL6
 
 * Fri Dec 25 2009  Scott Cantor  <cantor.2@osu.edu>  - 2.4-1
